protobuf_serialization/libp2p_varint.nim

```diff
@@ -0,0 +1,392 @@
+  ## Maximum encoded length of 32bit integer is 5 octets.
+  when sizeof(outval) == 8:
+    var value: uint64 =
+      if int64(outval) < 0'i64:
+        not(uint64(outval) shl 1)
+      else:
+        uint64(outval) shl 1
+  else:
+    var value: uint32 =
+      if int32(outval) < 0'i32:
+        not(uint32(outval) shl 1)
+      else:
+        uint32(outval) shl 1
+  result = PB.putUVarint(pbytes, outsize, value)
+
+proc putSVarint*(pbytes: var openarray[byte], outsize: var int,
+                 outval: PBSomeSVarint): VarintStatus {.inline.} =
+  ## Encode signed integer ``outval`` (``int32`` or ``int64``) and store it to
+  ## array ``pbytes``.
+  ##
+  ## On success ``outlen`` will hold number of bytes (octets) used to encode
+  ## unsigned integer ``v``.
+  ##
+  ## If there not enough bytes available in buffer ``pbytes``, ``Incomplete``
+  ## error will be returned and ``outlen`` will be set to number of bytes
+  ## required.
+  ##
+  ## Maximum encoded length of 64bit integer is 10 octets.
+  ## Maximum encoded length of 32bit integer is 5 octets.
+  when sizeof(outval) == 8:
+    result = PB.putUVarint(pbytes, outsize, uint64(outval))
+  else:
+    result = PB.putUVarint(pbytes, outsize, uint32(outval))
+
+template varintFatal(msg) =
+  const m = msg
+  {.fatal: m.}
+
+proc putVarint*[T: PB|LP](vtype: typedesc[T], pbytes: var openarray[byte],
+                nbytes: var int, value: SomeVarint): VarintStatus {.inline.} =
+  when vtype is PB:
+    when (type(value) is PBSomeSVarint) or (type(value) is PBZigVarint):
+      result = putSVarint(pbytes, nbytes, value)
+    elif (type(value) is PBSomeUVarint):
+      result = PB.putUVarint(pbytes, nbytes, value)
+    else:
+      varintFatal("Protobuf's varint do not support type [" &
+                  typetraits.name(type(value)) & "]")
+  elif vtype is LP:
+    when (type(value) is LPSomeVarint):
+      result = LP.putUVarint(pbytes, nbytes, value)
+    else:
+      varintFatal("LibP2P's varint do not support type [" &
+                   typetraits.name(type(value)) & "]")
+
+proc getVarint*[T: PB|LP](vtype: typedesc[T], pbytes: openarray[byte],
+                          nbytes: var int,
+                          value: var SomeVarint): VarintStatus {.inline.} =
+  when vtype is PB:
+    when (type(value) is PBSomeSVarint) or (type(value) is PBZigVarint):
+      result = getSVarint(pbytes, nbytes, value)
+    elif (type(value) is PBSomeUVarint):
+      result = PB.getUVarint(pbytes, nbytes, value)
+    else:
+      varintFatal("Protobuf's varint do not support type [" &
+                  typetraits.name(type(value)) & "]")
+  elif vtype is LP:
+    when (type(value) is LPSomeVarint):
+      result = LP.getUVarint(pbytes, nbytes, value)
+    else:
+      varintFatal("LibP2P's varint do not support type [" &
+                   typetraits.name(type(value)) & "]")
+
+proc encodeVarint*(vtype: typedesc[PB],
+                   value: PBSomeVarint): seq[byte] {.inline.} =
```

Comment:
This can write directly to an output stream. This way, it can be asked to populate a regular array on the stack with `unsafeMemoryOutput`, or it can write to an arbitrary output stream without allocating a small sequence and copying its bytes over.